// views/m/user/UserCenter/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import * as tdom from "@po-to/tomato-jquery";
import UserCenterNav = require('views/m/user/UserCenterNav');
import userInfo = require('views/m/user/UserInfo');
import wallet = require('views/m/user/Wallet');
import valueAddService = require('views/m/user/ValueAddService');

class VPresenter extends project.VPresenter {
    /**模块 */
    private UserCenter_vue: any;
    private user;
    private _aside: JQuery;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        var that = this;
        var vueDom = this.find("#UserCenter_vue").get(0);
        var user = this.user = model.globalData.user;
        let UserCenter_vue: any = this.UserCenter_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                openService: false,
                len: model.globalData.supplier_vas
            },
            watch: {
                'user.phone': "checkInfo",
                code: "checkInfo",
            },
            methods: {
                /**编辑个人信息 */
                userInfo: function () {
                    tomato.getVPresenter<UserCenterNav>(funs.moduleToUrl('m/user/UserCenterNav'), (UserCenterNav) => {
                        UserCenterNav.setBody(this, "UserInfo");
                    })
                    tomato.getVPresenter<userInfo>(funs.moduleToUrl('m/user/UserInfo'), (userInfo) => {
                        (that.parent as tomato.Dialog).appendChild(userInfo);
                    })
                },
                /**查看钱包 */
                wallet: function () {
                    tomato.getVPresenter<UserCenterNav>(funs.moduleToUrl('m/user/UserCenterNav'), (UserCenterNav) => {
                        UserCenterNav.setBody(this, "Wallet");
                    })
                    tomato.getVPresenter<wallet>(funs.moduleToUrl('m/user/Wallet'), (wallet) => {
                        wallet.init();
                        (that.parent as tomato.Dialog).appendChild(wallet);
                    })
                },
                /**开通增值服务 */
                valueAddService: function () {
                    tomato.getVPresenter<UserCenterNav>(funs.moduleToUrl('m/user/UserCenterNav'), (UserCenterNav) => {
                        UserCenterNav.setBody(this, "ValueAddService");
                    })
                    tomato.getTopDialog().close();
                    tomato.getVPresenter<valueAddService>(funs.moduleToUrl('m/user/ValueAddService'), (valueAddService) => {
                        funs.loadPage(valueAddService);
                    })
                },
                /**注销登录 */
                logOut: function () {
                    api.Logout(function () {
                        //success
                        window.location.href = '/index.html';
                    }, function () {
                        //fail
                        alert('注销登录失败，请重试')
                    })
                }
            }
        })
        this._watchEvent();
        //this._els = this._getElements();
        this._aside = $('<div style="width:54px;height:100%;" data-back="1"></div>');
        this._aside.on("click", function () {
            (that.parent as tomato.Dialog).close();
        })
    }
    init() {
        // if (model.globalData.supplier_vas.length > 0) {
        //     this.UserCenter_vue.openService = false;
        // } else {
        //     this.UserCenter_vue.openService = true;
        // }
    }
    getAside() {
        return this._aside;
    }
}
export = VPresenter;
