// views/m/user/UserCenterNav/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import * as tdom from "@po-to/tomato-jquery";
import userInfo = require('views/m/user/UserInfo');
import wallet = require('views/m/user/Wallet');
import valueAddService = require('views/m/user/ValueAddService');
import userCenter = require('views/m/user/UserCenter');

class VPresenter extends project.VPresenter {
    /**模块 */
    private UserCenterNav_vue: any;
    private user;
    private dialog;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        // this.dialog = this.body.parent;
        let user = this.user = model.globalData.user;
        let vueDom = this.find("#UserCenterNav_vue").get(0);
        let UserCenterNav_vue: any = this.UserCenterNav_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                isAvatar: true
            },
        })
        this._watchEvent();
        this._els = this._getElements();
    }
    /**更新用户头像 */
    public updateAvatar() {
        if ("" == this.user.avatar) {
            this.UserCenterNav_vue.isAvatar = true;
        } else {
            this.UserCenterNav_vue.isAvatar = false;
        }
    }
    private _evt_logout() {
        api.Logout(function () {
            //success
            window.location.reload();
        }, function () {
            //fail
            funs.tip('注销登录失败，请重试', '', 'fail');
        })
    }
    /**回个人中心 */
    private _evt_userAvatar() {

        tomato.getVPresenter<userCenter>(funs.moduleToUrl('m/user/UserCenter'), (userCenter) => {
            tomato.getTopDialog().appendChild(userCenter);
        })
    }
    /**编辑个人信息 */
    private _evt_userInfo() {
        tomato.getVPresenter<userInfo>(funs.moduleToUrl('m/user/UserInfo'), (userInfo) => {
            tomato.getTopDialog().appendChild(userInfo);
        })
        this.setBody(this, "UserInfo");
    }
    /**查看钱包 */
    private _evt_wallet() {
        tomato.getVPresenter<wallet>(funs.moduleToUrl('m/user/Wallet'), (wallet) => {
            wallet.init();
            tomato.getTopDialog().appendChild(wallet);
        })
        this.setBody(this, "Wallet");
    }
    /**开通增值服务 */
    private _evt_valueAdd() {
        this.setBody(this, "ValueAddService");
        tomato.getTopDialog().close();
        tomato.getVPresenter<valueAddService>(funs.moduleToUrl('m/user/ValueAddService'), (valueAddService) => {
            funs.loadPage(valueAddService);
        })
    }
    /**注销登录 */
    private _evt_logOut() {
        api.Logout(function () {
            //success
            window.location.href='/Index';
        }, function () {
            //fail
            alert('注销登录失败，请重试')
        })
    }
    public setBody(body, type) {
        // this.body = body;
        this.find(".curTab").removeClass("curTab");
        if (type == "UserInfo") {
            this.find(".userInfo").addClass("curTab");
        } else if (type == "Wallet") {
            this.find(".wallet").addClass("curTab");
        } else if (type == "ValueAddService") {
            this.find(".valueAdd").addClass("curTab");
        }
    }
}
export = VPresenter;
