// views/m/common/Header/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import * as model from "views/global/common/Model";
import * as commonPage from "views/m/common/CommonPage";
import Menu = require('views/m/common/Menu');


class VPresenter extends project.VPresenter {
    /**模块 */
    public Header_vue: any;
    /**user */
    private user;
    /**Menu */
    private _menu: Menu;
    public menuDialog: tdom.CommonDialog;
    public rightDialog: tdom.CommonDialog;

    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        let vueDom = this.find("#Header_vue").get(0);
        let selected_province = model.globalData.selected_province;
        let user = this.user = model.globalData.user;
        let unread = this.user = model.globalData.unread_num;
        let Header_vue: any = this.Header_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                selected_province: selected_province,
                isActive: false,
                unread: unread
            },
            computed:{
                unreadCom:function(){
                    if((this as any).unread.private_unread >9){
                        return '…';
                    }
                    return (this as any).unread.private_unread;
                }
            }
        })
        this._watchEvent();

        this.menuDialog = new tdom.CommonDialog({
            className: "Dialog-Menu",
            effect: tdom.DialogEffect.slideDown,
            masked: true,
            size: { width: tomato.DialogSize.Content, height: tomato.DialogSize.Content },
            position: { x: tomato.DialogPosition.Center, y: tomato.DialogPosition.Top },
            offset: { x: 0, y: $(this.view[0]) }
        });
        this.menuDialog.close = function (): boolean {
            return tdom.CommonDialog.prototype.close.call(this, false);
        }
        this.rightDialog = new tdom.CommonDialog({
            className: "rightMenu",
            effect: tdom.DialogEffect.slideLeft,
            masked: true,
            size: { width: 300, height: tomato.DialogSize.Full },
            position: { x: tomato.DialogPosition.Right, y: tomato.DialogPosition.Top },
            bodyEffect: tdom.TurnEffect.slid,
            asideEffect: tdom.TurnEffect.slid,
        });
        this.rightDialog.close = function (): boolean {
            return tdom.CommonDialog.prototype.close.call(this, false);
        }
        tomato.application.appendChild(this.menuDialog);
        tomato.application.appendChild(this.rightDialog);
    }
    public init() {
        return tomato.getVPresenter<Menu>(funs.moduleToUrl('m/common/Menu'), (Menu) => {
            Menu.addListener(tomato.DialogEvent.Closed, () => {
                this.Header_vue.isActive = false;
            });
            Menu.addListener(tomato.DialogEvent.Focused, () => {
                this.Header_vue.isActive = true;
            })
            this._menu = Menu;
            this.menuDialog.appendChild(Menu);
        });
    }
    /**打开菜单 */
    _evt_showMenu(data: any, target: HTMLElement, hit: HTMLElement) {
        if (this.menuDialog.state == tomato.DialogState.Focused) {
            this.menuDialog.close();
        } else {
            this.menuDialog.focus();
        }
    }
    /**打开用户中心 */
    private _evt_showUserCenter() {
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/user/UserCenter'), (userCenter) => {
            this.rightDialog.appendChild(userCenter);
            this.rightDialog.focus();
        })
    }
    /**打开自己的项目 */
    private _evt_showJoinedProjects() {
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/project/JoinedProjects'), (JoinedProjects) => {
            funs.loadPage(JoinedProjects);
        })
    }
}
export = VPresenter;
