// nodejs/global/model/demand.ts

export interface supplier {
    id: string,
    nickname: string,
    company_name: string,
    avatar: string,
    status: number,
    online: number,
    partic_at: number,
}
export interface demandConfigItem {
    name: string,
    id: string,
}
export interface demandConfig {
    typeOptions: {
        0: any,
        1: any,
        2: any,
        3: any,
        4: any,
        5: any,
    },
    locationOptions: {
        1: demandConfigItem,
        2: demandConfigItem,
        3: demandConfigItem,
    },
    colorOptions: {
        1: demandConfigItem,
        2: demandConfigItem,
        3: demandConfigItem,
    },
    spanOptions: {
        1: demandConfigItem,
        2: demandConfigItem,
        3: demandConfigItem,
        4: demandConfigItem,
    }
}

export interface user {
    id: string,
    nick: string,
    sex: number,
    phone: string
}
export interface DemandInfo {
    user: user,
    type_name: string,
    type: number,
    cate:string,
    cate_name:string,
    location_name: string,
    location: number,
    color_name: string,
    color: number,
    malf:string,
    malf_name:string,
    span: string,
    span_id: number,
    address: string,
    size: string,
    budget: number,
    id: string, //项目ID；
    uid:string,
    prov:string,
    city: string,
    image: string,	//需求场地图片
    status: string, //10进行中，60: 交易结束
    intm: number,
    uptm: number,
    note: string,
    audio: string, //客服和客户之间的录音链接
    bid_fee: number, //抢单服务费
    bid_cnt: number,
    supplier_partic_status:string,
    supplier_list: supplier[],
    budgetFormat: {budget:string,unit:string},
    lastMsg?:any,
    im_unread?:any
}
export interface DemandList{
    [key:string]:DemandInfo
}
export interface DemandListByType{
    title:string,
    list: {double:DemandInfo[]}[]
}
