// views/m/project/ProjectList/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as api from "views/global/common/API";
import * as model from "views/global/common/Model";
import * as demand from 'nodejs/global/model/demand';
import DemandDetail = require('views/m/DemandDetail');
import * as Vue from 'vue';

class VPresenter extends project.VPresenter {
    
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._watchEvent();
        // 特殊处理 台湾省 ， 更新列表
        if (model.globalData.user.prov == "台湾省") {
            model.globalData.selected_province.province_id = "台湾省";
            model.globalData.selected_province.province_name = "台湾省";
            var lastTime = new Date().getTime() +'';
            api.GetDemandList({"sort_val":lastTime},function(data){
                model.globalData.project_list.length = 0;
                model.globalData.project_list.push(...data);
            },function(){
                // 不处理
            })  
        }
        new Vue({
            el: $("#ProjectList-container")[0],
            data: {items: model.globalData.project_list},
            methods: {
                updateTime : function(startTime){
                        startTime = startTime + '';
                        var text;
                        // var stime = new Date(startTime).getTime();//转换
                        var stime = startTime * 1000;
                        var date = new Date().getTime();
                        var gapMin = Math.floor((date - stime)/1000/60);
                        if(gapMin <= 0){
                            //防止偶尔出现的js取时间不准确
                            gapMin = 1;
                        }
                        var gapHour = Math.floor((date - stime)/1000/60/60);

                        if(gapHour <= 24){
                            text = gapHour + '小时前';
                            if(gapMin < 60){
                                text = gapMin + '分钟前';
                            }
                        }else if(gapHour <= 24*30){
                            gapHour = Math.floor(gapHour/24);
                            text = gapHour + '天前'
                        }else if(gapHour <= 24*30*12){
                            gapHour = Math.floor(gapHour/24/30);
                            text = gapHour + '个月前'
                        }else{
                            //防止意外情况，暂时如此处理
                            text = '刚刚'
                        }
                        return text;
                },
                iJoined : function(list:any[]){
                    return list.some(function(item){
                        return item.id == model.globalData.user.user_id;
                    })
                }
            },
            
        })
        
    }
    //项目的点击事件
    _evt_projectDetail(id:string) {
        var pro_info = this.getProjectData(id);
        if(pro_info){
            model.globalData.current_demand = pro_info;
            var flag = model.globalData.demand.id == model.globalData.current_demand.id;
            model.globalData.currIsSelfDemand = flag;
            tomato.getVPresenter<DemandDetail>(funs.moduleToUrl('m/DemandDetail'), (DemandDetail) => {
                //弹出详情页
                if(DemandDetail.vpid.indexOf('demand=') >0){
                    DemandDetail.vpid = DemandDetail.vpid.replace(/\?(demand=\d*)?/,'?demand='+model.globalData.current_demand.id);
                }else{
                    DemandDetail.vpid += '?demand='+model.globalData.current_demand.id;
                }
                funs.loadPage(DemandDetail)
            })
        }
    }
    //获取单个项目的详细信息
    getProjectData(did:string) : demand.DemandInfo|undefined{
        for (let a of model.globalData.project_list) {
            if (a.id == did) {
                return a
            }
        }
    }
    
    
}
export = VPresenter;
