// views/m/user/ValueAddService/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import * as tdom from "@po-to/tomato-jquery";
import SMSPopup = require('views/m/user/SMSPopup');
import ProvinceSelector = require('views/m/common/ProvinceSelector');


class VPresenter extends project.VPresenter {
    /**模块 */
    public ValueAddService_vue: any;
    /**用户信息 */
    private user;
    /**supplier_vas */
    private _provList;
    /**修改第几个省份 */
    private _index: number;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        let height = this.find(".city-pay").height();
        let that = this;

        let user = this.user = model.globalData.user;
        // let selectedProv = model.globalData.selected_province;
        let provList = this._provList = model.globalData.supplier_vas;// = new Array();
        let vueDom = this.find("#ValueAddService_vue").get(0);
        let ValueAddService_vue: any = window['aaa'] = this.ValueAddService_vue = new Vue({
            el: vueDom,
            data: {
                user: user,
                supplier_vas: provList,
                selectedProv: "广东省",
                btnMessage: '立即开通',
                isOpen: true,//true:默认未开通，false:有开通
                isActive: false//false:不展开

            },
            computed: {

            },
            methods: {
                /**开通弹窗 */
                sImmediatelyOpen: function (event) {
                    // that._index = parseInt($(event.currentTarget).parents('.city-parent').find(".index").text());
                    let months = parseInt(event.target.getAttribute('value'));
                    let money: number;
                    let price: number;
                    switch (months) {
                        case 1: money = 59; months = 2; price = 59; break;
                        case 6: money = 50 * 6;price = 50; break;
                        default: money = -1;
                    }
                    let prov = $(event.currentTarget).parents('.city-parent').find(".city-name").text();
                    //判断值
                    let len = that._provList.length;
                    for (let i = 0; i < len; i++) {
                        if (prov == that._provList[i].prov) {
                            that._index = i;
                            break;
                        } else {
                            that._index = NaN;
                        }
                    }
                    //先调用方法,打开弹窗
                    tomato.getVPresenter<SMSPopup>(funs.moduleToUrl('m/user/SMSPopup'), (SMSPopup) => {
                        SMSPopup.setMoney(price, months, money, prov, that._index);
                        tdom.open(SMSPopup, tdom.DialogTarget.Blank, {
                            size: { width: '90%', height: tomato.DialogSize.Content },
                            effect: tdom.DialogEffect.scale,
                            masked: true
                        });
                    })
                },
                /**展开 */
                showPrice: function (event) {
                    let prov = $(event.currentTarget).find(".city-name").text();
                    console.log(event.target);

                    let curHeight = $(event.currentTarget).parent().find(".city-pay").height();
                    let autoHeight = $(event.currentTarget).parent().find(".city-pay").css('height', 'auto').height();
                    if (curHeight <= height) {
                        //收回
                        ValueAddService_vue.isActive = false;
                        $(event.currentTarget).parent().find(".city-pay").slideUp();
                        $(event.currentTarget).css('border-bottom', "1px solid #e3e8ee");
                        $(event.currentTarget).find(".city-name").css("color", "#83848e");
                        $(event.currentTarget).find(".icon-circle-down").removeClass("rate");

                    } else {
                        //展开
                        ValueAddService_vue.isActive = true;
                        $(event.currentTarget).parent().find(".city-pay").slideDown();
                        $(event.currentTarget).css('border-bottom', "none");
                        $(event.currentTarget).find(".city-name").css("color", "#212022");
                        $(event.currentTarget).find(".icon-circle-down").addClass("rate");

                    }
                },
                /**选择省份 */
                chooseProvince: function () {
                    if (!ValueAddService_vue.isOpen) {
                        $(".city-add").removeClass("hide");
                        $(".other-city").addClass("hide");
                    }
                    tomato.getVPresenter<ProvinceSelector>(funs.moduleToUrl('m/common/ProvinceSelector'), (ProvinceSelector) => {
                        ProvinceSelector.setProv("SMS");
                        funs.loadPage(ProvinceSelector);
                    })
                },
                /**取消新增城市 */
                closeProvience: function () {
                    $(".other-city").removeClass("hide");
                    $(".city-add").addClass("hide");
                    that.ValueAddService_vue.isOpen = false;
                },
            }
        });
        this._watchEvent();
        //this._els = this._getElements();
    }
    _afterInstallTo() {
        setTimeout(function () {
            // 如果当前页面是支付宝跳转过来的
            var href = window.location.href;
            if (href.indexOf('fromAlipay') > 0) {
                var records = localStorage.getItem('openSmsRecord');
                if (records) {
                    var recordsArgs = JSON.parse(records);
                    tomato.getVPresenter<SMSPopup>(funs.moduleToUrl('m/user/SMSPopup'), (SMSPopup) => {
                        SMSPopup._evt_ensure(recordsArgs);
                    });
                }
            }
        }, 1)
    }
    public init() {
        this._provList = model.globalData.supplier_vas;
        if (this._provList.length > 0) {
            this.ValueAddService_vue.isOpen = false;
            let len = this._provList.length;
            var newArr: any = [];
            let end_date: string, btnMessage: string;
            for (let i = 0; i < len; i++) {
                if (101 == this._provList[i].type) {
                    let str = this._provList[i].end_date.replace(/\./g, "-");
                    end_date = str.replace(/-/g, ".");
                    // alert("str:"+str);
                    var timestamp1 = new Date(str.replace(/-/g, "/") + ' 23:59:59');
                    // alert("timestamp1:"+timestamp1);
                    let timestamp2 = new Date();//当前时间
                    // alert("timestamp2:"+timestamp2);
                    var d = timestamp1.getTime() - timestamp2.getTime();
                    this._provList[i].end_date = end_date;

                    if (d > 0) {
                        //未到期
                        btnMessage = '续费';
                        this._provList[i].dateClass = "normal";
                        this._provList[i].btnMessage = btnMessage;
                    } else if (d <= 0) {
                        //已到期
                        btnMessage = '立即开通'
                        this._provList[i].dateClass = "expire";
                        this._provList[i].btnMessage = btnMessage;
                    } else {
                        btnMessage = "报错了";
                        this._provList[i].dateClass = "normal";
                        this._provList[i].btnMessage = btnMessage;
                    }
                    newArr.push(this._provList[i]);
                }
            }
            this._provList = newArr;
            this.ValueAddService_vue.supplier_vas = this._provList;
        } else {
            this.ValueAddService_vue.isOpen = true;
        }
    }
    public setProv(Prov_name) {
        this.ValueAddService_vue.selectedProv = Prov_name;
        let len = this.ValueAddService_vue.supplier_vas.length;
        for (let i = 0; i < len; i++) {
            if (this.ValueAddService_vue.selectedProv == this.ValueAddService_vue.supplier_vas[i].prov) {
                this.ValueAddService_vue.btnMessage = "续费";
                break;
            }
            this.ValueAddService_vue.btnMessage = "立即开通";

        }
    }
}
export = VPresenter;
